accepted/0000-caasp-nodes.md

```diff
@@ -0,0 +1,277 @@
+- other functions that Uyuni/SUSE Manager needs to achieve all other operations
+
+Whitelisting `state.apply` for getting the package list also allows the forbidden operations described above that make use of `state.apply`.
+If we do not whitelist `state.apply`, then Uyuni/SUSE Manager cannot:
+
+- Check and display the patch level of the CaaS Platform node
+- Apply Salt states
+
+We have two options:
+
+- CaaS Platform module in Salt
+- Minion blackout disable/enable
+
+##### CaaS Platform module in Salt
+
+If Uyuni/SUSE Manage uses a different function name to invoke package operations on a CaaS Platform node rather than relying on `state.apply`, then minion blackout is a solution.
+In fact, as the Salt whitelisting algorithm filters by the function name, the white list would be:
+
+- `test.ping`
+- `caasp.info_installed`: returns the installed packages on the node
+- `caasp.pkginstall`: install a package on the node
+- other functions that Uyuni/SUSE Manager needs to achieve all other operations
+
+All other functions are in blackout mode.
+
+This solutions needs:
+- Modifications in Uyuni/SUSE Manager [SaltServerActionService](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/suse/manager/webui/services/SaltServerActionService.java): if the targeted minion is a `caasp_node` system type, switch the corresponding Salt operation to use the `caasp.*` Salt module
+- Implementation (and upstream) of a Salt module that lives under the `caasp.*` namespace
+
+Value of this solution:
+* It can work with plain Salt
+* It protects from `pkg.install` (and others) usage against the forbidden operations
+* Requires code changes in Uyuni/SUSE Manager and Salt
+* This workflow does not constitute a radical change for a user
+
+##### Minion blackout disable/enable
+
+One step back: the minion blackout white list only contains:
+
+- `test.ping`
+
+The idea would be that when the user needs to issue any action on the CaaS Platform node from Uyuni/SUSE Manager, there is the possibility to:
+
+1. Temporarily disable minion blackout from Uyuni/SUSE Manager using a Salt formula.
+2. The user then executes any action that was intended for the CaaS Platform node.
+3. Finally, minion blackout is enabled again:
+    - Manually by the user
+    - Automatically, as a hook after every action issued by the user
+    - Automatically, in a timely fashion (e.g. after 10 minutes)
+
+The functionality would be used also for troubleshooting and installing troubleshooting tools and issuing commands.
+
+The risk of breaking the cluster by issuing any forbidden operation is still present while the blackout is disabled.
+
+Value of this solution:
+* Easy and quick to implement
+* The user can still break its cluster by issuing any forbidden action while the cluster is not in the blackout
+* The Workflow for the user is different and mechanical, can lead to errors
+
```

Comment:
[disable_modules/whitelist-modules](https://docs.saltstack.com/en/latest/ref/configuration/minion.html#std:conf_minion-disable_modules) could be another alternative to blackout.

Reply:
What you are suggesting is the following: whitelist the `state` module and disable the other execution module. In other words: SUSE Manager will be able to issue package/patch operations but users from the Salt command line will not.

Do I get it right?

Reply:
Hmm and what would prevent users to still install packages via a state in that case?

Reply:
Ok didn't think it through. Now the problem that I see with [disable_modules/whitelist-modules ](https://docs.saltstack.com/en/latest/ref/configuration/minion.html#std:conf_minion-disable_modules) is that they are part of minion configuration file which means minion will need to be restarted every time we change something(update modules list) there.

While the `blackout` is more flexible in the sense that it can be switch on/off with just pillar update.


> Hmm and what would prevent users to still install packages via a state in that case?

For sles/opensuse, if  `zypperpkg` module is defined under `disable_modules` then user cannot install or perform any package related operation using `state` or directly calling the `pkg` module. 
Errors are not pretty but at least it works.I quickly execute a state on a minion to install a package where `zypperpkg` was in the list of `disabled_modules` and I got the following error.

```
  ID: test
    Function: pkg.installed
      Result: False
     Comment: State 'pkg.installed' was not found in SLS 'teststate'
              Reason: 'pkg' __virtual__ returned False
     Changes:   
```

Reply:
In addition to the difference that @admd has just explained, the difference between blackout and whitelist is that blacklist allows to granularly whitelist functions, while whitelist of modules enables all the functions in a module.